### PR TITLE
ci(nix): use explicit versions for clang and gcc

### LINKED ISSUE #1934: Package ‘libcxx-13.0.0’ in /nix/store/.../llvm/13/libcxx/default.nix:42 is marked as broken, refusing to evaluate
**Describe the bug**
`nix-shell` fails with an error about a broken llvm package.

**Firmware version and revision**
Master branch at d448d7b8416f2ee985ed078f482bf05ac49214c3 with up to date submodules.

**Desktop/smartphone setup (please complete the following information):**
 - OS: macOS Monterey 12.0.1
 - NixOS 2.4

**To Reproduce**

- `git submodule update --init --recursive --force`
- `poetry shell`
- `nix shell`

### PR DESCRIPTION
**Pull request by prusnak**
Supersedes https://github.com/trezor/trezor-firmware/pull/1968

Fixes #1934

### CODE REVIEW

**Comment by mmilata** [comment]
Looks good, pulls clang-format 12 for me. I'd also use the same llvmPackages for libclang (`LIBCLANG_PATH` attribute definition) so that we don't have to download two versions of LLVM.

**Comment by prusnak** [comment]
> Looks good, pulls clang-format 12 for me. I'd also use the same llvmPackages for libclang (`LIBCLANG_PATH` attribute definition) so that we don't have to download two versions of LLVM.

Ah, good catch. PR reworked in cba8bf273326cb929fb6b8cb398ef23f58629d6e

File: ci/shell.nix [hunk 1]
```diff
@@ -37,6 +37,8 @@ let
     # to use official binary, remove rustfmt from buildInputs and add it to extensions:
     extensions = [ "clippy" ];
   };
+  gcc = nixpkgs.gcc11;
+  llvmPackages = nixpkgs.llvmPackages_12;
 in
 with nixpkgs;
```
Comment (mmilata) [inline ci/shell.nix:43]: > bindings introduced by with do not shadow bindings introduced by other means
didn't know this

**Review by mmilata — APPROVED** [review]
LGTM

However this is weird:
```
% cd trezor-firmware
% nix-shell
$ gcc --version
gcc (GCC) 10.3.0
```

10.3 is the version of gcc-arm-embedded :thinking: 

It happens also in master so was not introduced in this PR.

**Comment by prusnak** [comment]
```
$ gcc --version
gcc (GCC) 10.3.0
```

What does `which gcc` say?

**Comment by mmilata** [comment]
Ok, this is unrelated to gcc-arm-embedded as it also happens outside our repo:
```
% nix-shell -p gcc11

[nix-shell:~]$ gcc --version
gcc (GCC) 10.3.0

[nix-shell:~]$ which gcc
/nix/store/da9l6ifizmh883j94gzx5n51h8lrrp00-gcc-wrapper-10.3.0/bin/gcc
```
I'm on NixOS unstable channel.

**Comment by prusnak** [comment]
> It happens also in master so was not introduced in this PR.

On master this is legit, because gcc points to gcc10 on Linux and gcc10 is on 10.3.0

**Comment by prusnak** [comment]
@mmilata what you see is exactly the same issue I see here on darwin: https://github.com/trezor/trezor-firmware/pull/1968#issuecomment-985420969

The gcc installed via Linux stdenv trumps gcc installed in nix-shell, the same as llvm installed via Darwin stdenv trumps llvm installed in nix-shell.

**Comment by mmilata** [comment]
Huh, that's weird o_O. There has to be some way to use custom gcc version on linux though, I suppose we're not the first to want it.

**Comment by prusnak** [comment]
> I suppose we're not the first to want it.

The fix is trivial :-) We need to use stdenvNoCC: f727691348aa22e815a1d9fdc7ef77838792f19d

**Review by mmilata — APPROVED** [review]
That kinda makes sense, yeah:D

**Pull request merged**